tests/subsys/net/lib/nrf_cloud/cloud/src/main.c

```diff
@@ -0,0 +1,1880 @@
+
+	zassert_ok(ret,	"return should be 0 when success");
+	zassert_equal(STATE_INITIALIZED, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the initialized state after disconnected successfully");
+}
+
+/* TEST NRF_CLOUD_SHADOW_DEVICE_STATUS_UPDATE */
+
+/* Verify that the function will fail if the
+ * internal state of the lib is not dc_connected
+ */
+ZTEST(nrf_cloud_test, test_shadow_device_status_update_not_dc_connected)
+{
+	/* Data for shadow device status update */
+	struct nrf_cloud_device_status dev_status_t = {
+		.modem = NULL,
+		.svc = NULL
+	};
+
+	/* Custom fakes for nrf_cloud_send */
+	nct_cc_send_fake.custom_fake = fake_nct_cc_send__succeeds;
+	nct_dc_stream_fake.custom_fake = fake_nct_dc_stream__succeeds;
+	nct_dc_send_fake.custom_fake = fake_nct_dc_send__succeeds;
+	nct_dc_bulk_send_fake.custom_fake = fake_nct_dc_bulk_send__succeeds;
+	/* Custom fakes for shadow device status update */
+	nrf_cloud_device_status_encode_fake.custom_fake = fake_device_status_encode__succeeds;
+
+	/* Cloud is in idle state */
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	int ret = nrf_cloud_shadow_device_status_update(&dev_status_t);
+
+	zassert_equal(-EACCES, ret, "return should be -EACCES when not in dc_connected state");
+
+	/* Cloud is initialized */
+	init_cloud_success();
+
+	ret = nrf_cloud_shadow_device_status_update(&dev_status_t);
+	zassert_equal(-EACCES, ret, "return should be -EACCES when not in dc_connected state");
+
+	/* Cloud is connected */
+	connect_cloud_success();
+	ret = nrf_cloud_shadow_device_status_update(&dev_status_t);
+	zassert_equal(-EACCES, ret, "return should be -EACCES when not in dc_connected state");
+}
+
+/* Verify that the function will fail if the
+ * cloud device status encode function fails
+ */
+ZTEST(nrf_cloud_test, test_shadow_device_status_update_status_encode_failed)
+{
+	/* Data for shadow device status update */
+	struct nrf_cloud_device_status dev_status_t = {
+		.modem = NULL,
+		.svc = NULL
+	};
+
+	/* Custom fakes for nrf_cloud_send */
+	nct_cc_send_fake.custom_fake = fake_nct_cc_send__succeeds;
+	nct_dc_stream_fake.custom_fake = fake_nct_dc_stream__succeeds;
+	nct_dc_send_fake.custom_fake = fake_nct_dc_send__succeeds;
+	nct_dc_bulk_send_fake.custom_fake = fake_nct_dc_bulk_send__succeeds;
+	/* Custom fakes for shadow device status update */
+	nrf_cloud_device_status_encode_fake.custom_fake = fake_device_status_encode__fails;
+
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	/* Connect to the cloud with dc_connect */
+	connect_cloud_dc_success();
+
+	int ret = nrf_cloud_shadow_device_status_update(&dev_status_t);
+
+	zassert_not_equal(0, ret, "return should not be 0 when status encoding fails");
+}
```

Comment:
This is definitely a test that would do better with MQTT mocking. You could actually have MQTT return an error code and incorporate the nct's natural response to that in the test.